go-controller/pkg/clustermanager/network_cluster_controller.go

```diff
@@ -543,17 +543,37 @@ func newIPAllocatorForNetwork(netInfo util.NetInfo) (subnet.Allocator, error) {
 
 	subnets := netInfo.Subnets()
 	ipNets := make([]*net.IPNet, 0, len(subnets))
+	excludeSubnets := netInfo.ExcludeSubnets()
 	for _, subnet := range subnets {
 		ipNets = append(ipNets, subnet.CIDR)
+		if isLayer2UserDefinedPrimaryNetwork(netInfo) {
+			excludeSubnets = append(
+				excludeSubnets,
+				autoExcludeCIDRs(subnet.CIDR)...,
+			)
+		}
 	}
 
 	if err := ipAllocator.AddOrUpdateSubnet(
 		netInfo.GetNetworkName(),
 		ipNets,
-		netInfo.ExcludeSubnets()...,
+		excludeSubnets...,
 	); err != nil {
 		return nil, err
 	}
 
 	return ipAllocator, nil
 }
+
+func isLayer2UserDefinedPrimaryNetwork(netInfo util.NetInfo) bool {
+	return netInfo.IsPrimaryNetwork() && netInfo.TopologyType() == types.Layer2Topology
+}
+
```

Comment:
nit: we lost some good comments from your reverted commit wanna bring that back here?